OSDOCS-20033: Kueue 1.4 and DRA (#113996)
openshift/openshift-docs: modules/kueue-dra-resourceclaimtemplates.adoc

// Module included in the following assemblies:
//
// * ai_workloads/kueue/kueue-dra-integrating-dynamic-resource-allocation.adoc

:_mod-docs-content-type: PROCEDURE
[id="kueue-dra-resourceclaimtemplates_{context}"]
= Configuring the `ResourceClaimTemplate` path 

[role="_abstract"]

> You need to create a `ResourceClaimTemplate` path that users submit workloads that explicitly reference `ResourceClaimTemplate` objects. To do this, configure `deviceClassMappings` and add the DRA resource to your `ClusterQueue`. 

Reviewer: ```suggestion
You can configure {kueue-name} to manage quota for workloads that explicitly reference `ResourceClaimTemplate` objects. This requires configuring `deviceClassMappings` in the {kueue-name} custom resource (CR) and adding the DRA resource to your `ClusterQueue`.
```

Author: changed.

> .Procedure

Reviewer: Should we add `.Prerequisites` before this?

Author: No.

+
Multiple device classes can map to the same logical resource name. For example, if you have separate device classes for different GPU models but want a single quota pool, as shown in the following example: 
+
[source,yaml]
----
resources:
  deviceClassMappings:
  - name: nvidia.com/gpu
    deviceClassNames:
    - gpu-a100.nvidia.com
    - gpu-h100.nvidia.com
----

. Create a file called `rct-queues.yaml` that contains the following content:
+
.Example quota configuration for `ResourceClaimTemplate`
[source,yaml]
----
apiVersion: kueue.x-k8s.io/v1beta2
kind: ResourceFlavor
metadata:
  name: "default-flavor"
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  namespaceSelector: {}
  resourceGroups:
  - coveredResources: ["cpu", "memory", "nvidia.com/gpu"]
    flavors:
    - name: "default-flavor"
      resources:
      - name: "cpu"
        nominalQuota: 40
      - name: "memory"
        nominalQuota: 200Gi
      - name: "nvidia.com/gpu"
        nominalQuota: 8
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: LocalQueue
metadata:
  namespace: "default"
  name: "user-queue"
spec:
  clusterQueue: "cluster-queue"
----

> 

Reviewer 2: We're creating a file called `rct-queues.yaml` but never applying it. 

Suggestion: add a step to apply it

```
oc apply -f rct-queues.yaml
```

Author: Added this step.

. Create the workload by running the following command:
+
[source,terminal]
----

> $ oc create -f rct-job.yaml

Reviewer: I see the ordering issue here. Maybe something like ` Create a `ResourceClaimTemplate` and a workload to verify the configuration. Create a file called `rct-job.yaml` with the following content:`

Author: Updated. @sohankunkerkar Please verify.

> . Create a `ResourceClaimTemplate` and a workload to verify the configuration.  Create a file called `rct-job.yaml` by running the following command:

Author: fixed.

kind: ResourceClaimTemplate
metadata:
  name: my-gpu
  namespace: default
spec:
  spec:
    devices:
      requests:
      - name: gpu
        exactly:
          deviceClassName: gpu.nvidia.com 
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: rct-test-job-
  namespace: default
  labels:
    kueue.x-k8s.io/queue-name: user-queue 
spec:
  template:
    spec:
      restartPolicy: Never
      resourceClaims:
      - name: gpu
        resourceClaimTemplateName: my-gpu 
      containers:
      - name: worker
        image: registry.k8s.io/e2e-test-images/agnhost:2.53
        args: ["pause"]
        resources:
          claims:
          - name: gpu 
          requests:
            cpu: "1"
            memory: "200Mi"
----

`deviceClassName:`:: References the DeviceClass configured in deviceClassMappings.
`kueue.x-k8s.io/queue-name:`:: Identifies the local queue to submit the job to.
`resourceClaimTemplateName:`:: References the ResourceClaimTemplate defined above. The template must exist in the same namespace as the job.
`name:`:: Attaches the resource claim to this container.

.Verification

. Verify that the workload has been created and admitted:
+
[source,terminal]
----

> $ oc -n default get workloads

Reviewer 2: If we execute the steps like this, the workloads are not going to be created because the namespace "default" is not managed by Kueue.

So, we can either:
1. Add to the section (below) "If the workload is not admitted, verify the following" to check if the namespace is managed by Kueue
```
oc label namespace default kueue.openshift.io/managed=true
```
2. Or, add as a step asking the user to label the namespace, as we did in some other examples.

Author: Added the command (1.)

// Module included in the following assemblies:
//
// * ai_workloads/kueue/kueue-dra-integrating-dynamic-resource-allocation.adoc

:_mod-docs-content-type: PROCEDURE
[id="kueue-dra-resourceclaimtemplates_{context}"]

> = Configuring the `ResourceClaimTemplate` path 

Reviewer 3: From the OCP docs guidelines: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings

Do not use backticks or other markup in headings.
```suggestion
= Configuring the ResourceClaimTemplate path 
```

[role="_abstract"]

> You can configure {kueue-name} to manage quota for workloads that explicitly reference `ResourceClaimTemplate` objects. This requires configuring `deviceClassMappings` in the {kueue-name} custom resource (CR) and adding the DRA resource to your `ClusterQueue`.

Reviewer 3: Be sure to add the appropriate noun after each object reference.

.Procedure

. Use the following command to add a `deviceClassMappings` entry to the {kueue-name} configuration that maps each `DeviceClass` to a logical resource name for quota:
+
[source,yaml]  
----
$ oc patch kueue cluster -n openshift-kueue-operator --type=merge -p '{
  "spec": {
    "config": {
      "resources": {
        "deviceClassMappings": [{
          "name": "nvidia.com/gpu",
          "deviceClassNames": ["gpu.nvidia.com"]
        }]
      }
    }
  }
}'
----

> 

Reviewer 3: ```suggestion
+
where:
+
```
See https://redhat-documentation.github.io/supplementary-style-guide/#explain-commands-variables-in-code-blocks

`name`:: The resource name used in `ClusterQueue` quotas and `Workload` status.
`deviceClassNames`:: One or more `DeviceClass` names that map to this resource.
+
Multiple device classes can map to the same logical resource name. For example, if you have separate device classes for different GPU models but want a single quota pool, as shown in the following example: 
+
[source,yaml]
----
resources:
  deviceClassMappings:
  - name: nvidia.com/gpu
    deviceClassNames:
    - gpu-a100.nvidia.com
    - gpu-h100.nvidia.com
----

. Create a file called `rct-queues.yaml` that contains the following content:
+

> .Example quota configuration for `ResourceClaimTemplate`

Reviewer 3: ```suggestion
.Example quota configuration for the `ResourceClaimTemplate` object
```

[source,yaml]
----
apiVersion: kueue.x-k8s.io/v1beta2
kind: ResourceFlavor
metadata:
  name: "default-flavor"
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  namespaceSelector: {}
  resourceGroups:
  - coveredResources: ["cpu", "memory", "nvidia.com/gpu"]
    flavors:
    - name: "default-flavor"
      resources:
      - name: "cpu"
        nominalQuota: 40
      - name: "memory"
        nominalQuota: 200Gi
      - name: "nvidia.com/gpu"
        nominalQuota: 8
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: LocalQueue
metadata:
  namespace: "default"
  name: "user-queue"
spec:
  clusterQueue: "cluster-queue"
----

. Apply the `rct-queues.yaml` file:
+
[source,terminal]
----
$ oc apply -f rct-queues.yaml
----

> . Create a `ResourceClaimTemplate` and a workload to verify the configuration. Create a file called `rct-job.yaml` by running the following command:

Reviewer 3: ```suggestion
. Create a `ResourceClaimTemplate` object and a workload to verify the configuration. Create a file called `rct-job.yaml` by running the following command:
```

+
[source,terminal]
----
$ oc create -f rct-job.yaml
----
+
.Example `ResourceClaimTemplate` workload
[source,yaml]
----
apiVersion: resource.k8s.io/v1
kind: ResourceClaimTemplate
metadata:
  name: my-gpu
  namespace: default
spec:
  spec:
    devices:
      requests:
      - name: gpu
        exactly:
          deviceClassName: gpu.nvidia.com 
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: rct-test-job-
  namespace: default
  labels:
    kueue.x-k8s.io/queue-name: user-queue 
spec:
  template:
    spec:
      restartPolicy: Never
      resourceClaims:
      - name: gpu
        resourceClaimTemplateName: my-gpu 
      containers:
      - name: worker
        image: registry.k8s.io/e2e-test-images/agnhost:2.53
        args: ["pause"]
        resources:
          claims:
          - name: gpu 
          requests:
            cpu: "1"
            memory: "200Mi"
----

> 

Reviewer 3: ```suggestion
+
where:
+
```

> `deviceClassName:`:: References the DeviceClass configured in deviceClassMappings.

Reviewer 3: ```suggestion
`deviceClassName:`:: References the `DeviceClass` object configured in the `deviceClassMappings` entry.
```

`kueue.x-k8s.io/queue-name:`:: Identifies the local queue to submit the job to.

> `resourceClaimTemplateName:`:: References the ResourceClaimTemplate defined above. The template must exist in the same namespace as the job.

Reviewer 3: ```suggestion
`resourceClaimTemplateName:`:: References the `ResourceClaimTemplate` object defined above. The template must exist in the same namespace as the job.
```

`name:`:: Attaches the resource claim to this container.

.Verification

. Verify that the workload has been created and admitted:
+
[source,terminal]
----
$ oc -n default get workloads
----

. Verify that a `ResourceClaim` was created from the template:
+
[source,terminal]
----
$ oc -n default get resourceclaims
----
+
If the workload is not admitted, verify the following:
+
* Check if the namespace is managed by {kueue-name}:
+
[source,terminal]
----
$ oc label namespace default kueue.openshift.io/managed=true
----

* The `deviceClassMappings` in the Kueue CR maps the `DeviceClass` name to the resource name in coveredResources.
* The `ClusterQueue` has sufficient quota available.

> * The `ResourceClaimTemplate` exists in the same namespace as the job.

Reviewer 3: ```suggestion
* The `ResourceClaimTemplate` object exists in the same namespace as the job.
```

// Module included in the following assemblies:
//
// * ai_workloads/kueue/kueue-dra-integrating-dynamic-resource-allocation.adoc

:_mod-docs-content-type: PROCEDURE
[id="kueue-dra-resourceclaimtemplates_{context}"]
= Configuring the `ResourceClaimTemplate` path 

[role="_abstract"]
You need to create a `ResourceClaimTemplate` path that users submit workloads that explicitly reference `ResourceClaimTemplate` objects. To do this, configure `deviceClassMappings` and add the DRA resource to your `ClusterQueue`. 

.Procedure

> . Add a `deviceClassMappings`` entry to the {kueue-name} configuration that maps each `DeviceClass` to a logical resource name for quota, as shown in the following example:

Reviewer: ```suggestion
. Add a `deviceClassMappings` entry to the {kueue-name} configuration that maps each `DeviceClass` to a logical resource name for quota, as shown in the following example:
```

Author: Fixed.

+
[source,yaml]
----

> apiVersion: config.kueue.x-k8s.io/v1beta2

Reviewer 2: On OCP with the Kueue Operator, the user cannot create or apply a Configuration object directly. The operator owns it — it  generates the Configuration from the Kueue CR and writes it into a ConfigMap that the controller reads. If a user tries to oc apply that  YAML, there's no CRD for config.kueue.x-k8s.io/v1beta2 Configuration — it's not a Kubernetes resource you create, it's an embedded config format.

A possible way to set deviceClassMappings on OCP is through the Kueue CR:

```
  oc patch kueue cluster --type=merge -p '{
    "spec": {
      "config": {
        "resources": {
          "deviceClassMappings": [{
            "name": "nvidia.com/gpu",
            "deviceClassNames": ["gpu.nvidia.com"]
          }]
        }
      }
    }
  }'
```

Author: @sohankunkerkar ?

Reviewer: ```sh
// Module included in the following assemblies:
//
// * ai_workloads/kueue/kueue-dra-integrating-dynamic-resource-allocation.adoc

:_mod-docs-content-type: PROCEDURE
[id="kueue-dra-partitionable-devices_{context}"]
= Configuring partitionable devices

[role="_abstract"]
You can configure {kueue-name} to manage quota for partitionable devices based on actual device capacity rather than device count. Partitionable devices, such as NVIDIA Multi-Instance GPU (MIG) capable GPUs like the A100 or H100, allow a single GPU to be dynamically subdivided into smaller partitions.

When counter-based quota is configured, {kueue-name} charges quota in capacity units such as GPU memory rather than counting whole devices. For example, a `1g.5gb` MIG partition on an A100-40GB charges `4864Mi` of GPU memory quota, while a whole GPU charges `40320Mi`.

.Prerequisites

* You have cluster administrator permissions.
* You have installed {kueue-name} by using the {kueue-op}.
* You have created a `Kueue` CR.
* Your cluster is running {product-title} 4.22 or later.
* A DRA driver that publishes `consumesCounters` in `ResourceSlice` objects is installed, for example, `nvidia-dra-driver`.
* MIG is enabled on the GPU hardware.
* You have enabled the `DRAPartitionableDevices` Kubernetes feature gate by adding the `CustomNoUpgrade` feature set to the `FeatureGate` CR named `cluster`, as shown in the following example:
+
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  name: cluster
spec:
  featureSet: CustomNoUpgrade
  customNoUpgrade:
    enabled:
    - DRAPartitionableDevices
----
+
[WARNING]
====
Enabling the `CustomNoUpgrade` feature set on your cluster cannot be undone and prevents minor version updates. This feature set is not supported on production clusters. For information about enabling feature gates, see "Enabling features using feature gates".
====

.Procedure

. Verify that your DRA driver publishes counter data by running the following command:
+
[source,terminal]
----
$ oc get resourceslices -o jsonpath='{range .items[*]}{.spec.driver}{"\t"}{range .spec.devices[*]}{.name}: {.consumesCounters}{"\n"}{end}{end}'
----
+
.Example output
[source,terminal]
----
gpu.nvidia.com	gpu-0: [{"counterSet":"shared","counters":{"memory":{"value":"40Gi"}}}]
----
+
If the output does not show `consumesCounters` data, verify that your DRA driver version supports partitionable devices and that MIG is enabled on the GPU hardware.

. Configure counter-based quota by adding a `deviceClassMappings` entry with a `sources` section to the `config.resources` section of the {kueue-name} CR, as shown in the following example:
+
[source,yaml]
----
apiVersion: kueue.openshift.io/v1
kind: Kueue
metadata:
  name: cluster
  namespace: openshift-kueue-operator
spec:
  config:
    resources:
      deviceClassMappings:
      - name: gpu.memory # <1>
        deviceClassNames: # <2>
        - gpu.nvidia.com
        - mig.nvidia.com
        sources: # <3>
        - type: Counter
          counter:
            name: memory # <4>
            driver: gpu.nvidia.com
            deviceSelector: # <5>
              type: CEL
              cel:
                expression: "device.driver == 'gpu.nvidia.com'"
# ...
----
<1> The logical resource name used in `ClusterQueue` quotas. When counter-based sources are configured, quota is charged in capacity units rather than device count.
<2> The `DeviceClass` names that map to this resource. Include both the whole-GPU class (`gpu.nvidia.com`) and the MIG class (`mig.nvidia.com`).
<3> Defines how {kueue-name} computes the quota charge.
<4> The counter name must match a counter key published by the DRA driver in `ResourceSlice` devices.
<5> Scopes which devices are eligible for counter-based quota accounting.
+
[NOTE]
====
The {kueue-name} operator automatically enables the required {kueue-name} feature gates when it detects the `DRAPartitionableDevices` Kubernetes feature gate and `sources` are configured in `deviceClassMappings`. No manual {kueue-name} feature gate configuration is required.
====

. Create a `ClusterQueue` with counter-based quota. Set the quota in capacity units rather than device count. Create a file called `pd-queues.yaml` with the following content:
+
.Example quota configuration for partitionable devices
[source,yaml]
----
apiVersion: kueue.x-k8s.io/v1beta2
kind: ResourceFlavor
metadata:
  name: "default-flavor"
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  namespaceSelector: {}
  resourceGroups:
  - coveredResources: ["cpu", "memory", "gpu.memory"] # <1>
    flavors:
    - name: "default-flavor"
      resources:
      - name: "cpu"
        nominalQuota: 40
      - name: "memory"
        nominalQuota: 200Gi
      - name: "gpu.memory" # <2>
        nominalQuota: 800Gi
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: LocalQueue
metadata:
  namespace: "team-a"
  name: "user-queue"
spec:
  clusterQueue: "cluster-queue"
----
<1> The `gpu.memory` entry must match the `name` value in `deviceClassMappings`.
<2> Sets the total GPU memory quota. For example, `800Gi` accommodates twenty A100-40GB GPUs or equivalent MIG partitions.
+
[NOTE]
====
When `ClusterQueue` objects share a cohort, ensure all queues use the same unit scale for counter resources. {kueue-name} does not validate unit consistency across `ClusterQueue` objects.
====

. Apply the quota configuration by running the following command:
+
[source,terminal]
----
$ oc apply -f pd-queues.yaml
----

. Create a workload that requests a MIG partition. Create a file called `pd-job.yaml` with the following content:
+
.Example workload requesting a MIG partition
[source,yaml]
----
apiVersion: resource.k8s.io/v1
kind: ResourceClaimTemplate
metadata:
  namespace: team-a
  name: gpu-partition
spec:
  spec:
    devices:
      requests:
      - name: gpu
        exactly:
          deviceClassName: mig.nvidia.com # <1>
          count: 1
          selectors:
          - cel:
              expression: "device.attributes['gpu.nvidia.com'].profile == '1g.5gb'" # <2>
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: pd-test-job-
  namespace: team-a
  labels:
    kueue.x-k8s.io/queue-name: user-queue # <3>
spec:
  template:
    spec:
      containers:
      - name: worker
        image: registry.k8s.io/e2e-test-images/agnhost:2.53
        args: ["pause"]
        resources:
          claims:
          - name: gpu
          requests:
            cpu: "1"
            memory: "200Mi"
      resourceClaims:
      - name: gpu
        resourceClaimTemplateName: gpu-partition # <4>
      restartPolicy: Never
----
<1> References the MIG `DeviceClass`.
<2> Selects a specific MIG partition profile. Available profiles depend on the GPU model, for example, `1g.5gb`, `2g.10gb`, `3g.20gb`, or `7g.40gb` for the A100-40GB.
<3> Identifies the local queue to submit the job to.
<4> References the `ResourceClaimTemplate` defined above. The `ResourceClaimTemplate` must exist in the same namespace as the job.

. Create the workload by running the following command:
+
[source,terminal]
----
$ oc create -f pd-job.yaml
----

.Verification

. Verify that the workload is admitted and that quota was charged in capacity units by running the following command:
+
[source,terminal]
----
$ oc -n team-a get workloads -o jsonpath='{range .items[*]}{.metadata.name}: {.status.admission.podSetAssignments[0].resourceUsage}{"\n"}{end}'
----
+
.Example output
[source,terminal]
----
job-pd-test-job-xxxxx: {"cpu":"1","gpu.memory":"4864Mi","memory":"200Mi"}
----
+
The `gpu.memory` value reflects the actual memory capacity of the requested MIG partition rather than a device count of `1`.

. If the workload is not admitted, verify the following:
+
* The `DRAPartitionableDevices` Kubernetes feature gate is enabled on the cluster.
* The `deviceClassMappings` `name` value matches the resource name in `coveredResources`.
* The `counter.name` in `sources` matches a counter key in the `ResourceSlice` objects.
* The `ClusterQueue` has sufficient GPU memory quota for the requested partition size.
* MIG is enabled on the GPU hardware.

```

Author: Replaced the procedure with the above.

kind: Configuration
resources:
  deviceClassMappings:
  - name: example.com/gpu           # Logical resource name for quota
    deviceClassNames:
    - gpu.example.com               # DeviceClass name(s)
----

`name`:: The resource name used in `ClusterQueue` quotas and `Workload` status.
`deviceClassNames`:: One or more `DeviceClass` names that map to this resource.
+
Multiple device classes can map to the same logical resource name. For example, if you have separate device classes for different GPU models but want a single quota pool, as shown in the following example: 
+
[source,yaml]
----
resources:
  deviceClassMappings:
  - name: example.com/gpu
    deviceClassNames:
    - gpu-a100.example.com
    - gpu-h100.example.com
----

> . Use the followng command to add the DRA resource to your `ClusterQueue` by including the logical resource name from `deviceClassMappings` in the `coveredResources` of your `ClusterQueue`:

Reviewer: ```suggestion
. Use the following command to add the DRA resource to your `ClusterQueue` by including the logical resource name from `deviceClassMappings` in the `coveredResources` of your `ClusterQueue`:
```

Author: Fixed.

+
[source,terminal]
----

> $ oc apply -f https://kueue.sigs.k8s.io/examples/dra/sample-dra-queues.yaml

Reviewer 2: Instead of pointing to an external link, can we follow the same as https://github.com/openshift/openshift-docs/pull/113996/changes#diff-50e686ab942a05f0afecd2feb3d4f0e5f49175c442dbe9a013e83a970da6d9fcR49? (where cluster-queue.yaml file was created).

> $ oc apply -f https://kueue.sigs.k8s.io/examples/dra/sample-dra-queues.yaml

Reviewer 2: Instead of pointing to an external link, can we follow the same as https://github.com/openshift/openshift-docs/pull/113996/changes#diff-50e686ab942a05f0afecd2feb3d4f0e5f49175c442dbe9a013e83a970da6d9fcR49? (where cluster-queue.yaml file was created).

----
+
.Example output
[source,yaml]
----
apiVersion: kueue.x-k8s.io/v1beta2
kind: ResourceFlavor
metadata:
  name: "default-flavor"
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  namespaceSelector: {}
  resourceGroups:
  - coveredResources: ["cpu", "memory", "nvidia.com/gpu"]
    flavors:
    - name: "default-flavor"
      resources:
      - name: "cpu"
        nominalQuota: 40
      - name: "memory"
        nominalQuota: 200Gi
      - name: "nvidia.com/gpu"
        nominalQuota: 8
---
apiVersion: kueue.x-k8s.io/v1beta2
kind: LocalQueue
metadata:
  namespace: "default"
  name: "user-queue"
spec:
  clusterQueue: "cluster-queue"
----

> 

Reviewer 2: The procedure configures infrastructure (deviceClassMappings, ClusterQueue, LocalQueue) but never shows a workload YAML. A user finishes this procedure and doesn't know what a Job with a ResourceClaimTemplate looks like. The PD module includes a workload example — the RCT module could include it too.

Author: @sohankunkerkar ?

Reviewer 2: What clarification you need @StephenJamesSmith ?

Author: @anahas-redhat I met with Sohan last week and he said he would go through your comments that I had questions about (you were not able to attend the meeting). I put these comments (@sohankunkerkar ?) to indicate the comments he should address.

Reviewer 2: @StephenJamesSmith feel free to reschedule if you need clarification.
